### PR TITLE
Make Filters more intuitive and add Clear and Apply All Options

### LINKED ISSUE #279: Make UI for filters more intuitive
The selection of categories for the filters is not coherent.
One possible fix would be to "pre-select" all filters and let the user un-select them.

### PR DESCRIPTION
**Pull request by Armetiis**
### Description:
This PR improves the filter functionality in the ListScreen by making the filters more intuitive and adding "Clear All" and "Apply All" buttons. These changes enhance user experience and provide more control over filter options.

image

### Key Changes:
- Added "Clear All" and "Apply All" buttons to the main filter section.
- Implemented "Clear" and "Apply" buttons for each individual filter category (Protection, Rack Type, Capacity).
- Updated `ParkingViewModel` to support new clear and apply functions for each filter type and all filters collectively.
- Modified the initial state of filters to show all options by default.
- Adjusted the filter logic in the Parking View Model

### Testing:
- Unit tests have been added to verify the behavior of new filter functions in `ParkingViewModel`.
- UI tests have been updated to check the presence and functionality of new filter buttons.
- Manual testing has been performed to ensure smooth user interaction with the new filter options.

### Coverage
image

-- -
Closes #279

### CODE REVIEW

File: app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt [hunk 1]
```diff
@@ -293,12 +353,9 @@ class ParkingViewModel(
               TurfMeasurement.distance(_circleCenter.value!!, parking.location.center)
             }
             .filter { parking ->
-              (_selectedProtection.value.isEmpty() ||
-                  _selectedProtection.value.contains(parking.protection)) &&
-                  (selectedRackTypes.value.isEmpty() ||
-                      _selectedRackTypes.value.contains(parking.rackType)) &&
-                  (_selectedCapacities.value.isEmpty() ||
-                      _selectedCapacities.value.contains(parking.capacity)) &&
+              _selectedProtection.value.contains(parking.protection) &&
```
Comment (matttsss) [inline app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:356]: I just realized that this filter could be integrated with the third one and it could make the sorting faster if we have less elements in the list.

File: app/src/main/java/com/github/se/cyrcle/ui/list/ListScreen.kt [hunk 2]
```diff
@@ -269,26 +295,48 @@ fun FilterHeader(parkingViewModel: ParkingViewModel) {
...
+                          .padding(6.dp),
+                  testTag = title)
+              ClickableText(
+                  text = AnnotatedString(stringResource(R.string.list_screen_apply)),
+                  onClick = { onApply() },
+                  style =
+                      MaterialTheme.typography.labelSmall.copy(
+                          color = MaterialTheme.colorScheme.primary),
+                  modifier =
+                      Modifier.padding(horizontal = 4.dp).testTag("Apply" + title + "Button"))
```
Comment (matttsss) [inline app/src/main/java/com/github/se/cyrcle/ui/list/ListScreen.kt:335]: Here you could have used the string format `"Apply$titleButton"` instead of concatenation

File: app/src/test/java/com/github/se/cyrcle/model/parking/ParkingViewModelTest.kt [hunk 3]
```diff
@@ -155,4 +155,89 @@ class ParkingViewModelTest {
...
+    }
+    if (capacities) {
+      assert(parkingViewModel.selectedCapacities.value == ParkingCapacity.entries.toSet())
+    }
+  }
+
+  // Helper functions to assert the state of the filter options
+  private fun assertEmpty(protection: Boolean, rackTypes: Boolean, capacities: Boolean) {
+    if (protection) {
+      assert(parkingViewModel.selectedProtection.value == emptySet<ParkingProtection>())
```
Comment (matttsss) [inline app/src/test/java/com/github/se/cyrcle/model/parking/ParkingViewModelTest.kt:234]: Is there not a `.isEmpty()` methods for sets?

File: app/src/main/java/com/github/se/cyrcle/ui/list/ListScreen.kt (hunk 2 above)
Comment (Armetiis) [reply app/src/main/java/com/github/se/cyrcle/ui/list/ListScreen.kt:335]: You're right, I didn't overthink it as it was "just" a testTag. Turns out I underthinked it

File: app/src/test/java/com/github/se/cyrcle/model/parking/ParkingViewModelTest.kt (hunk 3 above)
Comment (Armetiis) [reply app/src/test/java/com/github/se/cyrcle/model/parking/ParkingViewModelTest.kt:234]: Mhhh i think you're right, i'll refactor this

File: app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt (hunk 1 above)
Comment (Armetiis) [reply app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:356]: Hi, I don't really understand what you mean by this. Could you maybe rephrase? Did you mean including this in the filter by distance with &&? I don't know how much we could benefit from it but I can do this, doesn't seem too hard. Is this also this that you meant when you said "unifying the filters"

Comment (matttsss) [reply app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:356]: Yes, sorry if I was unclear, but yes, the idea was to extend the predicate we have for the first filler

**Review by matttsss — APPROVED** [review]
All good for me, the changes didn't break anything and the UI is clean, thanks for the work :ok_hand:

File: app/src/main/java/com/github/se/cyrcle/ui/list/ListScreen.kt
Comment (andrepeiry) [inline app/src/main/java/com/github/se/cyrcle/ui/list/ListScreen.kt:1]: In this screen, everything looks nice, but I would note that all padding value (and others value) are hardcoded, I think a `smallPadding = 4.dp`, `mediumPadding = 8.dp` and `largePadding = 12.dp` could be a quick search and replace refactor.

**Review by andrepeiry — APPROVED** [review]
Except the small Magic number refactor, I think Matteo said it all.
After playing with the UI, I have found it nice and intuitive !

File: app/src/main/java/com/github/se/cyrcle/ui/list/ListScreen.kt
Comment (Armetiis) [reply app/src/main/java/com/github/se/cyrcle/ui/list/ListScreen.kt:1]: I believe we always did it like this for paddings and such values, unless there is some guideline that I am unaware of?
